### ISSUE TITLE
Intrinsics implementation of x86-64 SIMD algorithms

### ISSUE THREAD

**Issue by dcommander** [libjpeg-turbo/libjpeg-turbo#732]
There aren't any serious issues that demand this at the moment, but as compilers and CPUs evolve, new features are introduced that are difficult for us to implement at the assembly level and can lead to various problems if we don't implement them.  (Refer to #350, #707, #708, and #729.)  I would never attempt an intrinsics implementation of the i386 SIMD algorithms, because the lack of available registers would make such an implementation very difficult, if not impossible, to optimize.  (The i386 SIMD algorithms are basically a legacy feature anyhow.)  SSE2 still doesn't have a comfortable number of registers, when you consider that using xmm8-xmm15 can cause performance issues in some cases.  However, it might at least be worth attempting.  Just bear in mind that it would likely require hundreds of hours of labor to get this right, and an organization would have to be willing to fund 100% of that labor.  My experience with the Arm Neon intrinsics implementation was that certain compilers (GCC < v12) had poorly-optimized Neon intrinsics, so it was necessary to keep some of the assembly algorithms when using those compilers.  Such may be necessary in the near term with the proposed x86-64 intrinsics implementation as well, so it may similarly be a forward-looking feature that isn't always enabled by default.

I will not accept PRs for this feature, because:

1. I have more experience with the x86-64 libjpeg-turbo SIMD extensions than anyone else in the world, so this is one of those features whereby it would take me longer to clean up, optimize, test, document, and integrate an outside contribution than it would take me to implement the feature myself.
2. Since the feature would be "nice to have" rather than critical, my primary interest in it is frankly monetary.  (I do this for a living, not for a hobby.)

**Comment by dcommander** [comment]
I somewhat naively converted the SSE2 RGB-to-YCbCr color conversion module to intrinsics, and it performs slightly better than the hand-coded assembly.  That looks promising.

**Comment by dcommander** [comment]
Even with a less naive/much more readable implementation, GCC 4+ does a good job of register allocation as long as I keep the basic flow of the color conversion algorithm the same.  (NOTE: The color conversion algorithm tends to be the most register-intense of the algorithms.)  Thus, the readable intrinsics implementation performs as well as the naive intrinsics implementation.  The same is not always true of 32-bit code, however.  On a newer Intel CPU (3.6 GHz Xeon W-2123 circa 2018), the 32-bit readable intrinsics implementation performs within a few percent of the naive implementation and the hand-tuned NASM code.   On an older Intel CPU (2.8 GHz Xeon W3530 circa 2010), the 32-bit readable intrinsics implementation is about 15% slower than the naive implementation and the hand-tuned NASM code.

I might have been willing to extend the x86-64 intrinsics implementation to i386 if it required no additional effort, but such appears not to be the case.  Thus, I will continue under the assumption that i386 will retain the existing NASM code.  (It would have to do so for MMX and other legacy instruction sets anyway, because I'm definitely not signing up to reimplement those modules using intrinsics.)

**Comment by dcommander** [comment]
Unfortunately I have encountered a serious problem.  For whatever reason, Clang does a horrible job of translating x86-64 intrinsics into assembly.  Here's a simple example ([GodBolt link](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,selection:(endColumn:2,endLineNumber:9,positionColumn:2,positionLineNumber:9,selectionStartColumn:2,selectionStartLineNumber:9,startColumn:2,startLineNumber:9),source:'%23include+%3Cemmintrin.h%3E%0A%0A__m128i+test(__m128i+in1,+__m128i+in2)%0A%7B%0A++__m128i+tmp1,+tmp2,+tmp3%3B%0A++%0A++tmp1+%3D+_mm_slli_si128(in1,+8)%3B%0A++return+_mm_unpackhi_epi8(tmp1,+in2)%3B%0A%7D'),l:'5',n:'0',o:'C+source+%231',t:'0')),k:49.713825779640494,l:'4',m:100,n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:cclang_trunk,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'1',trim:'0',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,libs:!(),options:'-msse2+-O3',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+x86-64+clang+(trunk)+(Editor+%231)',t:'0')),header:(),k:50.28617422035952,l:'4',m:100,n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)):
```
#include <emmintrin.h>

__m128i test(__m128i in1, __m128i in2)
{
  __m128i tmp1, tmp2, tmp3;
  
  tmp1 = _mm_slli_si128(in1, 8);
  return _mm_unpackhi_epi8(tmp1, in2);
}
```
is translated into
```
test:
  pxor xmm2, xmm2
  punpckhbw xmm1, xmm2
  punpcklbw xmm0, xmm2
  movdqa xmm2, xmm0
  punpckhwd xmm2, xmm1
  punpcklwd xmm0, xmm1
  packuswb xmm0, xmm2
  ret
```
by the latest Clang release. Meanwhile, the latest GCC release translates it correctly:
```
"test":
  pslldq xmm0, 8
  punpckhbw xmm0, xmm1
  ret
```

This shows up as a 20-30% regression in color conversion performance when using Clang and intrinsics vs. NASM.  :(

If I specify `-msse4.2` rather than `-msse2`, then Clang translates the intrinsics into two instructions, but not the correct instructions:
```
test:
  pshufd xmm1, xmm1, 238
  punpcklbw xmm0, xmm1
  ret
```

It similarly mistranslates AVX2 intrinsics, but it translates them into the same number of instructions, so performance isn't affected ([GodBolt link](https://godbolt.org/z/nzx7f16e7)):

```
#include <immintrin.h>

 __m256i test(__m256i in1, __m256i in2)
{
  __m256i tmp1, tmp2, tmp3;
  
  tmp1 = _mm256_slli_si256(in1, 8);
  return _mm256_unpackhi_epi8(tmp1, in2);
}
```
Clang:
```
.LCPI0_0:
  .byte 0
  .byte 8
  .byte 1
  .byte 9
  .byte 2
  .byte 10
  .byte 3
  .byte 11
  .byte 4
  .byte 12
  .byte 5
  .byte 13
  .byte 6
  .byte 14
  .byte 7
  .byte 15
  .byte 0
  .byte 8
  .byte 1
  .byte 9
  .byte 2
  .byte 10
  .byte 3
  .byte 11
  .byte 4
  .byte 12
  .byte 5
  .byte 13
  .byte 6
  .byte 14
  .byte 7
  .byte 15
test:
  vpblendd ymm0, ymm0, ymm1, 204
  vpshufb ymm0, ymm0, ymmword ptr [rip + .LCPI0_0]
  ret
```
GCC:
```
"test":
  vpslldq ymm0, ymm0, 8
  vpunpckhbw ymm0, ymm0, ymm1
  ret
```

I have tried rewriting the algorithm to use shuffles instead of shifts.  That improves the situation, but the regression is still 15-25%.

Unless there is a way to goad Clang into using the correct instructions for particular intrinsics, then unfortunately this project is dead in the water.  It doesn't make sense to spend hundreds of hours tediously writing optimal intrinsics code if the compiler is going to second guess me and make it less optimal.  (NOTE: I have also tried all of the various `-O` settings, to no avail.)

**Comment by mr-c** [comment]
You could file a bug with llvm especially effective if you can include Compiler Explorer links demonstrating the problem .

**Comment by dcommander** [comment]
Referring to https://github.com/llvm/llvm-project/issues/159670, I am not very hopeful.  The specific issue preventing performance parity between assembly and intrinsics implementations of the SSE2 color conversion algorithm is probably fixable, since LLVM generates performance-equivalent instructions when targeting SSSE3 or above.  (There is no reason why it can't generate the same instructions with `-msse2`, since the generated instructions are SSE2 instructions.)  However, completely overhauling all of the x86-64 SIMD algorithms would require going through this same process 30 times.  I have spent more than 20 hours thus far re-developing the SSE2 and AVX2 color conversion algorithms and tracking down performance issues with same.  Granted that the color conversion algorithms are among the most complicated, which is why I started with them, but I am dubious that I have the budget to go through this process for all of the algorithms, especially if I encounter similar LLVM intrinsics translation issues with other algorithms (which is likely.)

I may have to give up on this for now.  :(

**Comment by dcommander** [comment]
Unfortunately, I have to put on my project manager hat and defer this feature.  I have been awarded a large grant through FLOSS/fund, but it needs to cover at least all of the [libjpeg-turbo 3.2 milestones](https://github.com/libjpeg-turbo/libjpeg-turbo/milestone/5.)  Given how long it took to implement the x86 color conversion algorithms with intrinsics, I am skeptical that enough funding will be available to finish this feature and all of the other 3.2 milestones.  The points in favor of this feature are:

- Improving code readability
    - However, I am mitigating that somewhat by overhauling the comments and formatting in the existing NASM code.
- Eliminating the need to modify the x86 SIMD code to support calling convention changes and other things that the compiler normally handles
    - However, all known issues in that regard have been fixed, and no new issues have been reported in two years.
- Eliminating the need to install NASM or YASM on a libjpeg-turbo build machine
    - However, that isn't a big deal, since binary builds of NASM/YASM are readily available for all supported platforms.
- Providing another set of SIMD intrinsics to act as a base for SIMDe
    - However, referring to #838, we were able to figure out how to make libjpeg-turbo's Arm/Neon SIMD modules perform decently with SIMDe, so experimenting with SIMDe and libjpeg-turbo is no longer blocking on this feature.

The points against this feature are:

- Hundreds of hours of labor with no visible improvement for end users
- The need to manage compiler revisions, as we did with the Arm/Neon SIMD extensions
    - I can't trust the translation of x86 intrinsics with clang, which means that I would need to benchmark each algorithm with minimally about a dozen platform permutations.  I would potentially need to file and track issues with LLVM for every algorithm as well.  In order to prevent performance regressions, it would probably be necessary to retain both assembly and intrinsics implementations of certain algorithms until all translation issues are fixed in clang and the fixes become readily available.  (To put numbers on this, the Arm/Neon SIMD extensions were overhauled with intrinsics five years ago, but I was only recently able to obsolete the assembly versions.)
 
I will make my work thus far available in a feature branch for reference.

Sorry I can't provide better news.  I am not saying "never" on this feature, but I am saying "not right now."

**Comment by dcommander** [comment]
In the interest of not letting good code go to waste, my AVX2 and SSE2 intrinsics implementations of color conversion have been pushed to the [dev.x86intrinsics](https://github.com/libjpeg-turbo/libjpeg-turbo/tree/dev.x86intrinsics) branch.  I updated the [LLVM issue](https://github.com/llvm/llvm-project/issues/159670) with instructions on how to benchmark the intrinsics vs. NASM versions of the color conversion algorithm.  I'll check back in on this once the rest of 3.2 is ready to go, which will likely be well into 2026.

**Comment by dcommander** [comment]
Cool.  The LLVM developers have already made a change that causes the compiler to behave the same way with `-msse2` as it does with `-mssse3`, compiling the intrinsics into `pshufd` and `punpcklbw` ([compiler explorer link](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,selection:(endColumn:2,endLineNumber:9,positionColumn:2,positionLineNumber:9,selectionStartColumn:2,selectionStartLineNumber:9,startColumn:2,startLineNumber:9),source:'%23include+%3Cemmintrin.h%3E%0A%0A__m128i+test(__m128i+in1,+__m128i+in2)%0A%7B%0A++__m128i+tmp1,+tmp2,+tmp3%3B%0A++%0A++tmp1+%3D+_mm_slli_si128(in1,+8)%3B%0A++return+_mm_unpackhi_epi8(tmp1,+in2)%3B%0A%7D'),l:'5',n:'0',o:'C+source+%231',t:'0')),k:49.713825779640494,l:'4',m:100,n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:cclang_trunk,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'1',trim:'0',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,libs:!(),options:'-msse2+-O3',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+x86-64+clang+(trunk)+(Editor+%231)',t:'0')),header:(),k:50.28617422035952,l:'4',m:100,n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)).

**Comment by dcommander** [comment]
Restoring the "funding needed" tag, because the FLOSS/fund grant we received last year is being exhausted by higher-priority features (including RISC-V vector support.)

**Comment by dcommander** [comment]
After further consideration, I am of the opinion that it doesn't make sense to spend hundreds of hours re-writing the existing SSE2 and AVX2 modules, since they work fine.  However, it may make sense to write a whole new set of modules using AVX-512 compiler intrinsics, since AVX-512 has access to 32 registers.  Basically, the NASM code would remain as a legacy fallback for older x86 CPUs, and the intrinsics code would be used with newer x86 CPUs.  That kills two birds with one stone:
 
1. It allows us to introduce a simpler and more readable x86 SIMD implementation into the libjpeg-turbo code base without going crazy trying to re-implement the SSE2 and AVX2 modules and work around various compiler issues.
2. It allows us to maximize performance on newer x86 hardware.  The speedup relative to AVX2 is still probably not going to be stellar, but I am OK with an incremental speedup if it gives us a more maintainable and forward-looking code base.

This would still require a lot of labor, but reducing the scope in this manner at least brings it into the realm of project management feasibility.

**Comment by arrowd** [comment]
> Unfortunately I have encountered a serious problem. For whatever reason, Clang does a horrible job of translating x86-64 intrinsics into assembly. Here's a simple example ([GodBolt link](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,selection:(endColumn:2,endLineNumber:9,positionColumn:2,positionLineNumber:9,selectionStartColumn:2,selectionStartLineNumber:9,startColumn:2,startLineNumber:9),source:'%23include+%3Cemmintrin.h%3E%0A%0A__m128i+test(__m128i+in1,+__m128i+in2)%0A%7B%0A++__m128i+tmp1,+tmp2,+tmp3%3B%0A++%0A++tmp1+%3D+_mm_slli_si128(in1,+8)%3B%0A++return+_mm_unpackhi_epi8(tmp1,+in2)%3B%0A%7D'),l:'5',n:'0',o:'C+source+%231',t:'0')),k:49.713825779640494,l:'4',m:100,n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:cclang_trunk,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'1',trim:'0',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:___c,libs:!(),options:'-msse2+-O3',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+x86-64+clang+(trunk)+(Editor+%231)',t:'0')),header:(),k:50.28617422035952,l:'4',m:100,n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)):

Just a note that this seems to be fixed in Clang 22.

**Comment by dcommander** [comment]
Thanks for the confirmation. That may or may not be relevant anymore, given the change in scope for this feature.
